### PR TITLE
Add multiple PV/PVC pair handling to persistent volume e2e test

### PR DESCRIPTION
**Pull request by copejon**
Adds the framework for creating, validating, and deleting groups of PVs and PVCs.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31732)

### CODE REVIEW

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: 1) if a nil test is necessary, see (2), why index the slice? Why not just test `if pvc != ni`l ?
2) Does the check for nil indicate that the pvcList can be sparse? And if so why isn't there a similar check for the pvList?

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: If the pvList is paired with the pvcList, meaning that for all pvs, pvList[i] is bound to pvcList[i], then this func could be a wrapper for waitAndValidateMultiplePVandPVC(). You would get the LogF and reuse code.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: maybe don't write "sets" which have a kubernetes definition. List? Slice?

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: prepend: "which are checked by the AfterEach ..."?

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: can this be a wrapper for completeTest() ?

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: You don't need `len(pvcs) > 0`, just let the for loop handle that.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Again, it looks like the pvcs slice is sparse due to the `pvc != nil`

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Here you check for pv != nil (again implying a sparse slice) but you didn't check for nil on L124.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Add comment here and in makePersistentVolume() that the reclaimPolicy is set to "Recycle" so that the nfs recycler is tested when the pvc is deleted.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: This func does not mirror makePersistentVolume()'s ability to include a ClaimRef in the pv. Since the claimRef logic is a bit convoluted  (two funcs deal with it) maybe it's better to yank the pvc arg from makePersistentVolume() and put all of the pre-bound claimRef logic in createPVCPV() ?

File: test/e2e/persistent_volumes.go
Comment (copejon) [inline test/e2e/persistent_volumes.go]: 1) When I've used `if pvc != nil.....` on line 126, the call returns after 1 iteration with remaining PVCs.  I haven't uncovered the reason for this yet but it performs as expected if I directly access the slice by index.
2) The nil check is there to skip deleted PVCs.  As the loop iterates, the call returns a nil value (as result of `deletePVCandValidate()` -> `pvc, err = c.PersistentVolumeClaims(ns).Get(pvc.Name)`.  Without it, an attempt to read pvc.UID causes a nil pointer error and stack trace.

Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Agreed re. sparse pvcList.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: why not just `fmt.Errorf("Failed to delete nil PV")` and delete line 67?

Comment (copejon) [inline test/e2e/persistent_volumes.go]: I can't think of why I changed that... Fixing

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Can pairCountIn be initialized as `var pairCountIn = len(pvPvcBindMap)`

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: There's a general problem of not cleaning up (ie, not deleting) pvs and pvc when the bindMap is returned as nil. Eg. what if you've deleted half of the pvcs then get a delete error? The Context's AfterEach will see the nil and skip it's delete of all of the bound pvs/pvcs. So the bindMap needs to be returned, but which version of the map? The passed-in arg? The partially created new map? A merge of both?

**Comment by copejon** [comment]
@saad-ali can you PTAL

cc @jeffvance

File: test/e2e/persistent_volumes.go
Comment (copejon) [inline test/e2e/persistent_volumes.go]: the PV is allowed to be cleaned up by the AfterEach.  We could explicitly delete them here although the code would be a bit redundant

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: How does this defer work when, just before the func returns, you nil the pv and pvc?

**Review by jeffvance — COMMENTED** [review]
May apply to all It()s

**Review by jeffvance — COMMENTED** [review]
LGTM.  @saad PTAL, thanks!

**Comment by jeffvance** [comment]
@saad-ali would you mind changing the size to S or M (if that maters), and can you do a quick review? It looks good to me.  Thanks.

**Comment by saad-ali** [comment]
@k8s-bot ok to test

**Comment by saad-ali** [comment]
Sorry for the delay guys. Planning on reviewing this tmr (Monday Sept 26).

File: test/e2e/persistent_volumes.go
Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: What is the primary purpose of maintaining this list? As far as I can tell, it's used primarily for cleanup.

It sounds like really you need two data maps, one for PVs and the other for PVCs that you created. And I think you might be able to get away with both mapping to bools (true if successfully bound)--you fetch the actual PV/PVC object as needed and can dereference the link as needed. The key in both maps would be the UID or namespace/name of the PV or PVC.

**Review by saad-ali — CHANGES_REQUESTED** [review]

File: test/e2e/persistent_volumes.go
Comment (copejon) [inline test/e2e/persistent_volumes.go]: In the current iteration we compare PVs by name or UID only.  Given that this pretty much guarantees stale data (problematic comparison as you point out), we'll be changing the keys to strings (PV.name)

File: test/e2e/persistent_volumes.go
Comment (copejon) [inline test/e2e/persistent_volumes.go]: I should have read this comment before writing the previous one.  You're right, I think using 2 maps of `string(name):bool(isbound)`, then Get()ing the objects as needed is cleaner and makes their purpose a bit more intuitive.

**Comment by copejon** [comment]
@saad-ali Thanks for the input!  I'll go ahead with the suggestions and ping you when we're ready for another look.

**Comment by saad-ali** [comment]
@copejon Let me know when this is ready for me to take a look at.

**Comment by jeffvance** [comment]
@eparis Can you take a look? I thought I filled out what was needed based on an email of yours. My work email is jvance@redhat.com. Thanks.

**Comment by eparis** [comment]
@jeffvance I can't do it for you any more :-(. I just invited you to the org, but you should be able to sign yourself up here: https://identity.linuxfoundation.org/projects/cncf

In any case, just the googlebot CLA is good enough right now, we aren't enforcing the CNCF CLA yet, but you need this fixed very soon.

**Comment by jeffvance** [comment]
@eparis I clicked the link you provided and checked the _Sign up to contribute to this project as an employee_ box, which then displayed this message:
```
The system was unable to associate you with any organization which has signed a Corporate Contributor License Agreement for this project. Please ensure you are using your organization email address on your profile, contact somebody within your organization who is responsible for overseeing authorized employee contributors for this project, or use the contact form below.
```
So, is there a lag between your invite and when the "system" recognizes my name, or do you need my github user (which is "jeffvance"), or should I sign up as an individual?  Or? Thanks!

**Comment by eparis** [comment]
@jeffvance I WILDLY guessing that the problem is that your linux foundation account it not using your redhat.com e-mail account. Please see if that is your problem. If not we'll need to talk to @caniszczyk and/or Eric Searcy (but I don't know his github id)

**Comment by copejon** [comment]
@saad-ali PTAL

**Comment by jeffvance** [comment]
@eparis @saad-ali I just updated my linux foundation profile to use my red hat email instead of my personal gmail. Can one of you please trigger a "re-try" for this pr?  Thanks!

**Comment by jeffvance** [comment]
@saad-ali @eparis Is there still a CLA issue for me? I've changed my profile to use my red hat email. Also, what's strange is that I've already done a few prs so the cla stuff was resolved in the recent past. Anyway, just a quick ping to ask you to take a look. Thanks.

File: test/e2e/persistent_volumes.go
Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: Consider using `NamespacedName` type from https://github.com/kubernetes/kubernetes/blob/master/pkg/types/namespacedname.go instead

Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: That'll allow you to get rid of `NS()` and `Name()` too

File: test/e2e/persistent_volumes.go
Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: If there is a failure cleaning up one PVC, will the remaining PVCs ever get cleaned up?

**Review by saad-ali — CHANGES_REQUESTED** [review]

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Great!

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: If the failure occurs in the normal path of deleting the claim to trigger the recycler (and to test that the related pv becomes Available) the remaining pvcs will be deleted in the AfterEach. However, if the failure occurs in the AfterEach itself, then I guess the remaining claims are not deleted. I am not certain how Expect() works inside an AfterEach. Do you know of a way to harden this code?

Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: @saad-ali I tested a failure in an AfterEach. In the log you'll see "Failure in Spec Teardown (AfterEach) ..." to indicate that the test failed inside an AfterEach. A failure in an AfterEach exits the AfterEach block and reports the failure. There can be a stack of AfterEach()s and thus a higher-level AfterEach may be invoked. For example, if a Context() AfterEach fails then the KubeDescribe AfterEach may be invoked. The final AfterEach is defined in the framework. This AfterEach deletes the namespace and this deletes the PVCs in that namespace. So, as long as all of the test claims live in the same ns they will be deleted, at least when the ns is deleted. However, our plan is to create pvcs and pod in several different namespaces, and in that case the claims residing in the other namespaces, other than the ns where the initial ns, will survive the test.  In fact, all the of the pvs (in the system global ns, remain after the test fails.

**Comment by jeffvance** [comment]
@saad-ali The GCE e2e failures above have nothing to do with this PR. Can you PTAL and add any additional suggestions for improvement, or add a LGTM so we can get this foundational PR merged? Thanks!

**Comment by jeffvance** [comment]
Ping @saad-ali ... thanks!

**Comment by jeffvance** [comment]
@saad-ali I think it's good now. My last change was to deal with cases where a deleted pvc results in the pv that had been bound to that claim being in phase Bound again.  PTAL, thanks.

**Comment by jeffvance** [comment]
@saad-ali Looking at the log for GCI GKE smoke e2e I see this error:
`E1018 10:14:28.795] FAIL: pull-kubernetes-e2e-gke-gci`
I assume for now that this error is independent from this pr but let me know if that's wrong. Thanks.

File: test/e2e/persistent_volumes.go
Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: Is this intentionally commented out?

File: test/e2e/persistent_volumes.go
Comment (saad-ali) [inline test/e2e/persistent_volumes.go]: Where is this method used?

**Review by saad-ali — CHANGES_REQUESTED** [review]
Almost there. Couple more comments.

These tests do not run in the Google CI at the moment. Do they run as part of Red Hat CI? I'm asking because I want to make sure they are passing and not-flaky.

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Good catch! That line was commented as part of my testing to see what happens if Expect() is called inside an AfterEach. Thanks!

File: test/e2e/persistent_volumes.go
Comment (jeffvance) [inline test/e2e/persistent_volumes.go]: Another good catch! This was code that was used at one point and eventually replaced.

**Comment by jeffvance** [comment]
@saad-ali Thank you for your review! I've pushed the latest to address your comments.

**Comment by jeffvance** [comment]
Did rebase.
@k8s-bot gci gke e2e test this

**Comment by jeffvance** [comment]
@saad-ali did the rebase. Are we good now?  Thanks!

**Comment by saad-ali** [comment]
Thanks. LGTM. Let's get his merged and then focus on getting these tests into a stable [non-flaky] state.

Looks like you need a rebase. Will add LGTM label after that.

**Comment by copejon** [comment]
@saad-ali Rebased!

**Comment by saad-ali** [comment]
Adding LGTM label.

Can you please run these tests to verify everything is working as intended before merge.

**Comment by copejon** [comment]
Re-tested against local and GCE cluster

**Comment by jeffvance** [comment]
Thanks @saad-ali ! So we're finally good to merge now?

**Pull request merged**
